### PR TITLE
HUB-175: User can hit pause page without session and be redirected to rp

### PR DESCRIPTION
**Pull request by rachelthecodesmith**
Add view for users without valid session but with the journey hint
cookie state to allow them to proceed to the RP service start page.

solo: @rachelthecodesmith

### CODE REVIEW

File: app/controllers/single_idp_journey_controller.rb [hunk 1]
```diff
@@ -44,9 +46,15 @@ def redirect_from_idp
...
     idp_entity_id = params['idpEntityId']
     uuid = params['singleIdpJourneyIdentifier'].to_s.downcase
-    rp_url = get_service_choice_url(transaction_id)
 
-    if !rp_url.nil? && valid_idp_choice?(idp_entity_id, transaction_id) && valid_uuid?(uuid)
+    rp_url = get_service_choice_url(get_service_list, transaction_id)
+
+    if rp_url.nil?
+      redirect_to verify_services_path
+      return
```
Comment (chrisclayson) [inline app/controllers/single_idp_journey_controller.rb]: Could this explicit `return` be avoided? Maybe with an `elsif` below?

Comment (rachelthecodesmith) [reply app/controllers/single_idp_journey_controller.rb]: It already has an elsif below that it used to be part of but I agree this is also a bit ugly. I just figured this at least broke us out of things as early as possible. Will have another look

File: app/controllers/partials/retrieve_federation_data_partial_controller.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
+module RetrieveFederationPartialController
```
Comment (jakubmiarka) [inline app/controllers/partials/retrieve_federation_data_partial_controller.rb]: the file is called retrieve_federation_*data*_partial_controller but the controller is called RetrieveFederationPartialController

File: app/controllers/paused_registration_controller.rb [hunk 3]
```diff
@@ -1,22 +1,25 @@
...
   layout 'slides', except: :index
 
   def index
-    if session_is_valid
-      set_transaction_from_session
-      @idp = IDENTITY_PROVIDER_DISPLAY_DECORATOR.decorate(selected_identity_provider)
-      render :with_user_session
+    if session_is_valid?
+      with_session
+    elsif journey_hint_cookie_state_present?
```
Comment (jakubmiarka) [inline app/controllers/paused_registration_controller.rb]: should this be true only if the last state is PENDING? we could then use `is_last_status?(status)` method and get rid off the `journey_hint_cookie_state_present?` below?

File: app/controllers/partials/retrieve_federation_data_partial_controller.rb (hunk 2 above)
Comment (chrisclayson) [reply app/controllers/partials/retrieve_federation_data_partial_controller.rb]: Well spotted! I missed that in my review!

File: spec/api_test_helper.rb [hunk 4]
```diff
@@ -37,10 +37,10 @@ def api_select_country_endpoint(session_id, country_code)
 
   def stub_transactions_list
     transactions = [
-        { 'simpleId' => 'test-rp', 'serviceHomepage' => 'http://localhost:50130/test-rp', 'loaList' => ['LEVEL_2'] },
-        { 'simpleId' => 'test-rp-noc3', 'serviceHomepage' => 'http://localhost:50130/test-rp-noc3', 'loaList' => ['LEVEL_2'] },
-        { 'simpleId' => 'headless-rp', 'serviceHomepage' => 'http://localhost:50130/headless-rp', 'loaList' => ['LEVEL_2'] },
-        { 'simpleId' => 'test-rp-with-continue-on-fail', 'serviceHomepage' => 'http://localhost:50130/test-rp-with-continue-on-fail', 'loaList' => ['LEVEL_2'] }
+        { 'simpleId' => 'test-rp', 'entityId' => 'http://www.test-rp.gov.uk/SAML2/MD', 'serviceHomepage' => 'http://localhost:50130/test-rp', 'loaList' => ['LEVEL_2'] },
```
Comment (jakubmiarka) [inline spec/api_test_helper.rb]: does the endpoint really return the entityId?

File: app/controllers/paused_registration_controller.rb [hunk 5]
```diff
@@ -34,21 +37,63 @@ def resume
...
+    @transaction = {
+        name: selected_rp,
+        homepage: get_rp_homepage(selected_rp)
+    }
+  end
+
+  def get_cookie_rp_from_service_list
+    last_rp_value = last_rp
+    return nil if last_rp_value.nil?
+    get_selected_rp_from_simple_id(get_service_list, last_rp_value)
```
Comment (jakubmiarka) [inline app/controllers/paused_registration_controller.rb]: the last_rp_value is an entityId if I'm not mistaken? not simpleId

File: app/controllers/paused_registration_controller.rb [hunk 6]
```diff
@@ -34,21 +37,63 @@ def resume
...
     @transaction = {
         name: current_transaction.rp_name,
         homepage: current_transaction_homepage
     }
   end
+
+  def set_transaction_from_cookie(selected_rp)
+    @transaction = {
+        name: selected_rp,
+        homepage: get_rp_homepage(selected_rp)
```
Comment (jakubmiarka) [inline app/controllers/paused_registration_controller.rb]: could we replace this method with `selected_rp.fetch('serviceHomepage', nil)` directly? that way we don't need the separate method to wich basically just wraps the fetch?

File: app/controllers/paused_registration_controller.rb [hunk 7]
```diff
@@ -34,21 +37,63 @@ def resume
...
   def set_transaction_from_session
     @transaction = {
         name: current_transaction.rp_name,
         homepage: current_transaction_homepage
     }
   end
+
+  def set_transaction_from_cookie(selected_rp)
+    @transaction = {
+        name: selected_rp,
```
Comment (jakubmiarka) [inline app/controllers/paused_registration_controller.rb]: does `selected_rp` represent an object? If so, should it be the `selected_rp.name`?

**Pull request merged**